types/react-router/index.d.ts

```diff
@@ -15,112 +15,82 @@
-  }
-  class MemoryRouter extends React.Component<MemoryRouterProps, void> {}
-
-
-  interface PromptProps {
-    message: string | ((location: H.Location) => void);
-    when?: boolean;
-  }
-  class Prompt extends React.Component<PromptProps, void> {}
-
-
-  interface RedirectProps {
-    to: H.LocationDescriptor;
-    push?: boolean;
-    from?: string;
-    path?: string;
-    exact?: boolean;
-    strict?: boolean;
-  }
-  class Redirect extends React.Component<RedirectProps, void> {}
-
-
-  interface RouteComponentProps<P> {
-    match: match<P>;
-    location: H.Location;
-    history: H.History;
-  }
-
-  interface RouteProps {
-    location?: H.Location;
-    component?: React.SFC<RouteComponentProps<any> | void> | React.ComponentClass<RouteComponentProps<any> | void>;
-    render?: (props: RouteComponentProps<any>) => React.ReactNode;
-    children?: (props: RouteComponentProps<any>) => React.ReactNode | React.ReactNode;
-    path?: string;
-    exact?: boolean;
-    strict?: boolean;
-  }
-  class Route extends React.Component<RouteProps, void> {}
-
-
-  interface RouterProps {
-    history: any;
-  }
-  class Router extends React.Component<RouterProps, void> {}
-
-
-  interface StaticRouterProps {
-    basename?: string;
-    location?: string | object;
-    context?: object;
-  }
-  class StaticRouter extends React.Component<StaticRouterProps, void> {}
-
-
-  interface SwitchProps extends RouteProps {
-  }
-  class Switch extends React.Component<SwitchProps, void> {}
+import * as React from 'react';
+import * as H from 'history';
+
+// This is the type of the context object that will be passed down to all children of
+// a `Router` component:
+export interface RouterChildContext<P> {
+  router: {
+    history: H.History
+    route: {
+      location: H.Location
+      match: match<P>
+    }
+  };
+}
+export interface MemoryRouterProps {
+  initialEntries?: H.LocationDescriptor[];
+  initialIndex?: number;
+  getUserConfirmation?: (() => void);
+  keyLength?: number;
+}
 
+export class MemoryRouter extends React.Component<MemoryRouterProps, void> {}
 
-  interface match<P> {
-    params: P;
-    isExact: boolean;
-    path: string;
-    url: string;
-  }
+export interface PromptProps {
+  message: string | ((location: H.Location) => void);
+  when?: boolean;
+}
+export class Prompt extends React.Component<PromptProps, void> {}
+
+export interface RedirectProps {
+  to: H.LocationDescriptor;
+  push?: boolean;
+  from?: string;
+  path?: string;
+  exact?: boolean;
+  strict?: boolean;
+}
+export class Redirect extends React.Component<RedirectProps, void> {}
 
+export interface RouteComponentProps<P> {
+  match: match<P>;
+  location: H.Location;
+  history: H.History;
+}
 
-  function matchPath<P>(pathname: string, props: RouteProps): match<P> | null;
+export interface RouteProps {
+  location?: H.Location;
+  component?: React.SFC<RouteComponentProps<any> | undefined> | React.ComponentClass<RouteComponentProps<any> | undefined>;
```

Comment:
This line wasn't modified by this PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16240/files#diff-bc5f906a3af0cde1a95a8a23cb993e96L68

The `react-router` definition needs some cleaning to do, @erkieh don't hesitate to send a new PR with the proposed changes.